src/AppInstallerCommonCore/Certificates.cpp

```diff
@@ -0,0 +1,511 @@
+            {
+                AICLI_LOG(Core, Verbose, << "Public key mismatch: Expected certificate [" << GetSimpleDisplayName(m_certificateContext.get()) << "], Actual certificate [" << GetSimpleDisplayName(certContext) << "]");
+                return false;
+            }
+        }
+
+        if (WI_IsFlagSet(m_pinning, PinningVerificationType::Subject))
+        {
+            if (!CertCompareCertificateName(
+                X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                &m_certificateContext.get()->pCertInfo->Subject,
+                &certContext->pCertInfo->Subject))
+            {
+                AICLI_LOG(Core, Verbose, << "Subject mismatch: Expected certificate [" << GetSimpleDisplayName(m_certificateContext.get()) << "], Actual certificate [" << GetSimpleDisplayName(certContext) << "]");
+                return false;
+            }
+        }
+
+        if (WI_IsFlagSet(m_pinning, PinningVerificationType::Issuer))
+        {
+            if (!CertCompareCertificateName(
+                X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                &m_certificateContext.get()->pCertInfo->Issuer,
+                &certContext->pCertInfo->Issuer))
+            {
+                AICLI_LOG(Core, Verbose, << "Issuer mismatch: Expected certificate [" << GetSimpleDisplayName(m_certificateContext.get()) << "], Actual certificate [" << GetSimpleDisplayName(certContext) << "]");
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    PinningChain::Node PinningChain::Node::Next()
+    {
+        if (!HasNext())
+        {
+            m_chain.get().emplace_back();
+        }
+
+        return { m_chain, m_index + 1 };
+    }
+
+    const PinningChain::Node PinningChain::Node::Next() const
+    {
+        THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), !HasNext());
+        return { m_chain, m_index + 1 };
+    }
+
+    void PinningChain::Node::RemoveNext()
+    {
+        m_chain.get().erase(m_chain.get().begin() + m_index + 1, m_chain.get().end());
+    }
+
+    bool PinningChain::Node::HasNext() const
+    {
+        return (m_index + 1 < m_chain.get().size());
+    }
+
+    PinningChain::Node::Node(std::vector<PinningDetails>& chain, size_t index) :
+        m_chain(chain), m_index(index) {}
+
+    PinningChain::Node PinningChain::Root()
+    {
+        if (m_chain.empty())
+        {
+            m_chain.emplace_back();
+        }
+
+        return { m_chain, 0 };
+    }
+
+    const PinningChain::Node PinningChain::Root() const
+    {
+        THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), m_chain.empty());
+        return { const_cast<std::vector<PinningDetails>&>(m_chain), 0 };
+    }
+
+    bool PinningChain::Validate(PCCERT_CHAIN_CONTEXT chainContext) const
+    {
+        if (m_chain.empty())
+        {
+            // An empty chain rejects all inputs.
+            AICLI_LOG(Core, Verbose, << "Empty pinning chain blindly rejecting chain context");

```

Comment:
>Verbose [](http://example.com/codeflow?start=28&length=7)

I think logging level of this function should be revisited?

Reply:
I changed this line specifically to `Warning` as an empty chain is not an expected use case.